### PR TITLE
use SecureRandom for number generator

### PR DESCRIPTION
**Pull request by wuboy0307**
I encounter this issue last week.
https://github.com/spree/spree/issues/6719

There are two things should be fixed.

using sample() with random seed may get into concurrency issue when two threads create number at the same time (at exactly the same millisecond may have the same random seed).
It would be better to use SecureRandom instead.

Also, according to this article https://makandracards.com/makandra/13901-understand-and-fix-race-conditions-with-duplicate-unique-keys-in-rails-mysql  
 I believe it's mandatory to add unique index to all "number" fields.

### CODE REVIEW

**Comment by damianlegawiec** [comment]
Hi @wuboy0307 - thank you for this contribution, can you fix hound issues (new ruby syntax) and failing specs? https://circleci.com/gh/spree/spree/5425

File: core/lib/spree/core/number_generator.rb [hunk 1]
```diff
@@ -35,17 +30,13 @@ def included(host)
...
+        candidate = @letters ? 36 : 10
+        number = loop do
+          random_token = SecureRandom.random_number(candidate**length).to_s(candidate).rjust(length, "0")
+          break random_token unless host.exists?(number: random_token)
         end
-      end
 
-      def new_candidate(length)
-        @prefix + length.times.map { @candidates.sample(random: @random) }.join
+        return @prefix + number
```
Comment (wuboy0307) [inline core/lib/spree/core/number_generator.rb]: I don't know why I got this issue. I didn't see any redundant return.

File: core/db/migrate/20170219132415_add_unique_index_to_number_fields.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class AddUniqueIndexToNumberFields < ActiveRecord::Migration
```
Comment (wuboy0307) [inline core/db/migrate/20170219132415_add_unique_index_to_number_fields.rb]: I don't know how to fix this either. Why do I need a comment?

Comment (damianlegawiec) [inline core/db/migrate/20170219132415_add_unique_index_to_number_fields.rb]: @wuboy0307 this is a hound bug, ignore it :)

**Comment by wuboy0307** [comment]
I don't know why circleci test failed. I didn't touch general_settings_controller.rb or orders_controller.rb. That must be some other issues in the version I fetch.

File: core/db/migrate/20170219132415_add_unique_index_to_number_fields.rb (hunk 2 above)
Comment (krzysiek1507) [inline core/db/migrate/20170219132415_add_unique_index_to_number_fields.rb]: @wuboy0307 @damianlegawiec It's a feature. ;) http://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/FrozenStringLiteralComment

File: core/lib/spree/core/number_generator.rb (hunk 1 above)
Comment (krzysiek1507) [inline core/lib/spree/core/number_generator.rb]: @wuboy0307 last line in method is a result so you don't need to write `return sth`

File: core/db/migrate/20170219132415_add_unique_index_to_number_fields.rb (hunk 2 above)
Comment (wuboy0307) [inline core/db/migrate/20170219132415_add_unique_index_to_number_fields.rb]: @krzysiek1507, So do I need to change anything?

File: core/lib/spree/core/number_generator.rb (hunk 1 above)
Comment (wuboy0307) [inline core/lib/spree/core/number_generator.rb]: sorry, java habit :P

**Comment by krzysiek1507** [comment]
@wuboy0307 check `$ ./build-ci.rb test (0)` section

File: core/db/migrate/20170219132415_add_unique_index_to_number_fields.rb (hunk 2 above)
Comment (krzysiek1507) [inline core/db/migrate/20170219132415_add_unique_index_to_number_fields.rb]: @damianlegawiec should we follow with frozen literals?

**Comment by wuboy0307** [comment]
@krzysiek1507  I am sorry but I have no idea what you are talking about. I am not quite familiar with spree testing. I am just someone who got into concurrency issue and try to fix it. As you can see, all I have modified are just a db migration and a number_generator.rb file, which have nothing to do with those tests... Can you guys help to figure it out?

**Comment by krzysiek1507** [comment]
@wuboy0307 it looks like your migration is incorrect:

```
Building: backend
Generating dummy Rails application...
Setting up dummy database...
DEPRECATION WARNING: Directly inheriting from ActiveRecord::Migration is deprecated. Please specify the Rails release the migration was written for:

  class AddUniqueIndexToNumberFields < ActiveRecord::Migration[4.2] (called from load at /home/ubuntu/spree/vendor/bundle/ruby/2.3.0/bin/rake:23)
rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

You must specify the index name
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/schema_statements.rb:762:in `index_name'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:846:in `block in method_missing'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:815:in `block in say_with_time'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:815:in `say_with_time'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:835:in `method_missing'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration/compatibility.rb:87:in `index_name_for_remove'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration/compatibility.rb:80:in `remove_index'
/home/ubuntu/spree/backend/spec/dummy/db/migrate/20170224116139_add_unique_index_to_number_fields.spree.rb:4:in `change'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:789:in `exec_migration'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:773:in `block (2 levels) in migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:772:in `block in migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:398:in `with_connection'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:771:in `migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration/compatibility.rb:121:in `migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:951:in `migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1212:in `block in execute_migration_in_transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1280:in `block in ddl_transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/database_statements.rb:232:in `block in transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/transaction.rb:189:in `within_new_transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/database_statements.rb:232:in `transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/transactions.rb:211:in `transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1280:in `ddl_transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1211:in `execute_migration_in_transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1184:in `block in migrate_without_lock'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1183:in `each'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1183:in `migrate_without_lock'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1132:in `block in migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1299:in `with_advisory_lock'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1132:in `migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1006:in `up'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:984:in `migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/tasks/database_tasks.rb:161:in `migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/railties/databases.rake:58:in `block (2 levels) in <top (required)>'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/rake-12.0.0/exe/rake:27:in `<top (required)>'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:74:in `load'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:74:in `kernel_load'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:27:in `run'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/cli.rb:332:in `exec'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/cli.rb:20:in `dispatch'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/cli.rb:11:in `start'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/exe/bundle:34:in `block in <top (required)>'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/friendly_errors.rb:100:in `with_friendly_errors'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/exe/bundle:26:in `<top (required)>'
/opt/circleci/.rvm/gems/ruby-2.3.1/bin/bundle:23:in `load'
/opt/circleci/.rvm/gems/ruby-2.3.1/bin/bundle:23:in `<main>'
ArgumentError: You must specify the index name
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/schema_statements.rb:762:in `index_name'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:846:in `block in method_missing'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:815:in `block in say_with_time'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:815:in `say_with_time'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:835:in `method_missing'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration/compatibility.rb:87:in `index_name_for_remove'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration/compatibility.rb:80:in `remove_index'
/home/ubuntu/spree/backend/spec/dummy/db/migrate/20170224116139_add_unique_index_to_number_fields.spree.rb:4:in `change'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:789:in `exec_migration'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:773:in `block (2 levels) in migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:772:in `block in migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:398:in `with_connection'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:771:in `migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration/compatibility.rb:121:in `migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:951:in `migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1212:in `block in execute_migration_in_transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1280:in `block in ddl_transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/database_statements.rb:232:in `block in transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/transaction.rb:189:in `within_new_transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/database_statements.rb:232:in `transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/transactions.rb:211:in `transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1280:in `ddl_transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1211:in `execute_migration_in_transaction'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1184:in `block in migrate_without_lock'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1183:in `each'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1183:in `migrate_without_lock'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1132:in `block in migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1299:in `with_advisory_lock'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1132:in `migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:1006:in `up'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/migration.rb:984:in `migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/tasks/database_tasks.rb:161:in `migrate'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/activerecord-5.0.1/lib/active_record/railties/databases.rake:58:in `block (2 levels) in <top (required)>'
/home/ubuntu/spree/vendor/bundle/ruby/2.3.0/gems/rake-12.0.0/exe/rake:27:in `<top (required)>'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:74:in `load'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:74:in `kernel_load'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/cli/exec.rb:27:in `run'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/cli.rb:332:in `exec'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/cli.rb:20:in `dispatch'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/cli.rb:11:in `start'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/exe/bundle:34:in `block in <top (required)>'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/lib/bundler/friendly_errors.rb:100:in `with_friendly_errors'
/opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.13.7/exe/bundle:26:in `<top (required)>'
/opt/circleci/.rvm/gems/ruby-2.3.1/bin/bundle:23:in `load'
/opt/circleci/.rvm/gems/ruby-2.3.1/bin/bundle:23:in `<main>'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
```

File: core/lib/spree/core/number_generator.rb [hunk 3]
```diff
@@ -35,17 +30,13 @@ def included(host)
...
 
-        loop do
-          candidate = new_candidate(length)
-          return candidate unless host.exists?(number: candidate)
-
-          # If over half of all possible options are taken add another digit.
-          length += 1 if host.count > Rational(BASE**length, 2)
+        candidate = @letters ? 36 : 10
+        number = loop do
+          random_token = SecureRandom.random_number(candidate**length).to_s(candidate).rjust(length, "0")
```
Comment (mauazua) [inline core/lib/spree/core/number_generator.rb]: please, use single-quotes here `'0'`

**Comment by mauazua** [comment]
@wuboy0307 please run specs on your local environment, because some of failing examples seem to do so because of changes introduced by you.

**Comment by jkelleyj** [comment]
Noticed that this is related to #7822

File: core/lib/spree/core/number_generator.rb [hunk 4]
```diff
@@ -1,18 +1,13 @@
...
     class NumberGenerator < Module
-      BASE           = 10
       DEFAULT_LENGTH = 9
-      NUMBERS        = (0..9).to_a.freeze
-      LETTERS        = ('A'..'Z').to_a.freeze
-
-      attr_accessor :prefix, :length
 
       def initialize(options)
         @random     = Random.new
```
Comment (jkelleyj) [inline core/lib/spree/core/number_generator.rb]: It looks like you're no longer using @random in this class with the secure random approach.

**Comment by jkelleyj** [comment]
I'm curious why we've decided to scrap the original approach (which may have been used and included in classes in its current form by repo users).  I was able to eliminate the collision issue by grabbing a new instance of random each time.  Might be a safer change for the community.

**Comment by wuboy0307** [comment]
@jkelleyj  For me, what is really important is to add unique index to protect the database record. For number generator, once collision issue can be fixed, any modification is okay for me. However, we already know that "the original approach" is buggy, why not just scrap it?  The number collision issue has been happened on our production environment. It caused our customer paying money on WRONG orders. It's unacceptable since it relate to money.  For me, what is safer is keeping our system stable and normal.

**Pull request closed without merging**